openedx/core/djangoapps/util/tests/test_ip.py

```diff
@@ -0,0 +1,194 @@
+
+        assert actual == expected
+
+        if warning_substr is None:
+            assert len(caught_warnings) == 0
+        else:
+            assert len(caught_warnings) == 1
+            assert warning_substr in caught_warnings[0]
+
+    @ddt.unpack
+    @ddt.data(
+        # No config
+        ({}, [], None, 0),
+        # Configured, but none of the headers are present
+        (
+            {},
+            [
+                {'name': 'CF-Connecting-IP', 'index': 0},
+                {'name': 'X-Forwarded-For', 'index': -2},
+                {'name': 'X-Real-IP', 'index': 0},
+            ],
+            None,
+            3,
+        ),
+        # One lookup failure before finding a usable header
+        (
+            {
+                'HTTP_X_FORWARDED_FOR': '1.1.1.1, 2.2.2.2, 3.3.3.3, 4.4.4.4'
+            },
+            [
+                {'name': 'CF-Connecting-IP', 'index': 0},
+                {'name': 'X-Forwarded-For', 'index': -2},
+                {'name': 'X-Real-IP', 'index': 0},
+            ],
+            '3.3.3.3',
+            1,
+        ),
+    )
+    def test_get_via_configured_headers(self, add_meta, cnf_headers, expected, warning_count):
+        self.request.META.update(add_meta)
+
+        with override_settings(**{'CLIENT_IP_HEADERS': cnf_headers}):
+            with warning_messages() as caught_warnings:
+                actual = ip.get_client_ip_via_configured_headers(self.request.META)
+
+        assert actual == expected
+        assert len(caught_warnings) == warning_count
+
+    @ddt.unpack
+    @ddt.data(
+        # Nothing usable
+        ([], None),
+        (['any-old-junk'], None),
+        # Simple cases, private and public
+        (['::1'], '::1'),
+        (['1.2.3.4'], '1.2.3.4'),
+        # If we get a public IP, don't worry about junk farther on
+        (['junk', '1.2.3.4', '2606:4700::'], '2606:4700::'),
+        (['junk', '2606:4700::', '10.0.0.1'], '2606:4700::'),
+        # Walk left until first public IP
+        (['1.2.3.4', '5.6.7.8', '10.0.0.1', '127.0.0.1'], '5.6.7.8'),
+        # Or until there's junk, even if the best we can do is a private IP
+        (['1.2.3.4', 'XXXXXXX', '10.0.0.1', '127.0.0.1'], '10.0.0.1'),
+    )
+    def test_conservative_walk(self, chain, expected):
+        assert ip.conservatively_pick_client_ip(chain) == expected
+
+    @ddt.unpack
+    @ddt.data(
+        # Publicly exposed server, perhaps
+        ({'REMOTE_ADDR': '4.3.2.1'}, '4.3.2.1'),
+        # ...maybe with someone trying to spoof their IP
+        ({'HTTP_X_FORWARDED_FOR': '5.5.5.5', 'REMOTE_ADDR': '4.3.2.1'}, '4.3.2.1'),
+        # If XFF is missing, just accept a private IP
+        ({'REMOTE_ADDR': '127.0.0.1'}, '127.0.0.1'),
+        # General case, walk from right
+        ({'HTTP_X_FORWARDED_FOR': '8.7.6.5, 4.3.2.1, 10.0.0.1', 'REMOTE_ADDR': '127.0.0.1'}, '4.3.2.1'),
+    )
+    def test_get_via_xff(self, add_meta, expected):
```

Comment:
1. Maybe I should just move all the tests to `test_get_via_xff` and drop the ones for `test_conservative_walk`, since the former is a thin wrapper around the latter. I don't like duplicating the tests unnecessarily.
2. Happy to change the order. I usually _write_ the exceptional cases first, but that doesn't mean they have to be first in source order!
3. Looks like it's possible for `REMOTE_ADDR` to be something other than an IP address if the request comes in over a UNIX domain socket rather than a TCP connection, but it should still be present. (I believe gunicorn provides an empty string.) I can make an adjustment to handle that properly in case we ever shift to use domain sockets for some reason.